src/auth/Crypto.cc

```diff
@@ -127,21 +122,32 @@ class HandlerImpl : public Handler {
   }
   int create(Random *random, bufferptr& secret) override;
   int validate_secret(const bufferptr& secret) override;
-  std::unique_ptr<KeyHandler> get_key_handler(const bufferptr& secret,
-                                              string& error) override;
+  std::unique_ptr<KeyHandler> get_key_handler(const bufferptr& secret) override;
 };
 
 #ifdef USE_NSS
 // when we say AES, we mean AES-128
 # define AES_KEY_LEN	16
 # define AES_BLOCK_LEN   16
 
-static int nss_aes_operation(CK_ATTRIBUTE_TYPE op,
-			     CK_MECHANISM_TYPE mechanism,
-			     PK11SymKey *key,
-			     SECItem *param,
-			     const bufferlist& in, bufferlist& out,
-			     std::string *error)
+class nss_exception : public std::exception {
```

Comment:
Could we implement these using std:error_code rather than just a wrapper on std::exception?

Reply:
we could add an `error_category` for nss error codes and provide mappings from those to strings. then we could either throw those as `system_error` exceptions, or avoid exceptions by having these interfaces take an `error_code&` instead of `std::string*`. that could be useful if any of our code was looking for specific error conditions. but the existing usage is basically just "did it fail? if so, log an error message" - and i think a custom exception type is a better fit for that, as it can provide some extra context like which api function returned that error code

what do you think?